plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/TypeClassPattern.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.jdbc.expression;
+
+import com.google.common.collect.ImmutableSet;
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.matching.Property;
+import io.trino.spi.type.Type;
+
+import java.util.Objects;
+import java.util.Set;
+
+import static java.util.Objects.requireNonNull;
```

Comment:
General comment - how about moving some of the classes (like this one) from `io.trino.plugin.jdbc.expression` into `trino-plugin-toolkit`? (can be done on a future PR)

Reply:
The DSL is geared towards JDBC connectors, since it produces `String` replacement. This is unlikely suitable output for non-JDBC connectors. 

Also, we will need to revisit the rewrites and make them more coupled with JDBC connectors, when we support projection pushdown with complex expressions. For example, we probably will need to carry also `JdbcTypeHandle`. cc @wendigo @hashhar 

@assaf2  What use-case do you have in mind?

Reply:
Since we chose to expose `ConnectorExpressionRewriter` under `trino-plugin-toolkit`, and since you're adding more capabilities, I thought that other connectors can potentially benefit those capabilities as well. IMO `TypeParameterPattern` and `TypePattern` class hierarchies are generic enough to be reused as-is by other connectors. Maybe we can add another hierarchy on top of `GenericRewrite` so any connector could write code like `.map("$modulus(left: integer_type, right: integer_type)").to(<something>)`

Reply:
> Since we chose to expose `ConnectorExpressionRewriter` under `trino-plugin-toolkit`,

yes, `ConnectorExpressionRewriter` is supposed to be generic just like `AggregateFunctionRewriter` is.
And in fact, `AggregateFunctionRewriter` is used in Pinot in a different way than it is used in JDBC connectors.

> TypeParameterPattern and TypePattern class hierarchies are generic enough to be reused as-is by other connectors.

yes, the `.map(...)` section is generic, as it defines a pattern over `ConnectorExpression`
the `GenericRewrite` is not generic in a sense of being applicable to different connector types.

We can move `ExpressionMappingParser` & co to plugin toolkit, if it helps anything. (i don't see the use-case yet)